lib/src/binding_mqtt/mqtt_client.dart

```diff
@@ -0,0 +1,245 @@
+import '../scripting_api/subscription.dart' as scripting_api;
+import 'mqtt_binding_exception.dart';
+import 'mqtt_config.dart';
+import 'mqtt_extensions.dart';
+import 'mqtt_subscription.dart';
+
+/// [ProtocolClient] for supporting the MQTT protocol.
+///
+/// Currently, only MQTT version 3.1.1 is supported.
+class MqttClient extends ProtocolClient {
+  /// Constructor.
+  MqttClient(
+    this._clientSecurityProvider,
+    MqttConfig? mqttConfig,
+  ) : _mqttConfig = mqttConfig ?? MqttConfig();
+
+  final MqttConfig _mqttConfig;
+
+  final ClientSecurityProvider? _clientSecurityProvider;
+
+  Future<BasicCredentials?> _obtainCredentials(
+    Uri uri,
+    Form? form, [
+    BasicCredentials? invalidCredentials,
+    bool unauthorized = false,
+  ]) async {
+    final requiresBasicAuthentication =
+        form?.requiresBasicAuthencation(invalidCredentials) ?? false;
+    final isDiscovery = form == null && unauthorized;
+
+    if (!(requiresBasicAuthentication || isDiscovery)) {
+      return null;
+    }
+
+    final basicCredentials = _clientSecurityProvider?.basicCredentialsCallback
+        ?.call(uri, form, invalidCredentials);
+
+    if (basicCredentials != null) {
+      return basicCredentials;
+    }
+
+    if (form != null) {
+      throw MqttBindingException(
+        'Form requires basic authencation but no credentials were provided.',
+      );
+    }
+
+    throw MqttBindingException(
+      'Discovery requires basic authencation but no credentials were provided.',
+    );
+  }
+
+  Future<MqttServerClient> _connectWithForm(Form form) async =>
+      _connect(form.resolvedHref, form);
+
+  Future<MqttServerClient> _connect(Uri brokerUri, Form? form) async {
+    final client = brokerUri.createClient();
+    final credentials = await _obtainCredentials(brokerUri, form);
+
+    MqttClientConnectionStatus? status;
+
+    try {
+      status = await client.connectWithCredentials(credentials);
+    } on NoConnectionException {
+      // Ask user for (new) credentials
+      final newCredentials =
+          await _obtainCredentials(brokerUri, form, credentials);
+      if (newCredentials != null) {
+        status = await client.connectWithCredentials(newCredentials);
+      }
+    }
+
+    if (status?.state != MqttConnectionState.connected) {
+      throw MqttBindingException('Connecting failed');
+    }
+
+    return client;
+  }
+
+  @override
+  Future<Content> invokeResource(Form form, Content content) async {
+    final client = await _connectWithForm(form);
+    final topic = form.topicName;
+    final qualityOfService =
+        form.qualityOfService ?? _mqttConfig.defaultActionQoS.mqttQoS;
+
+    client
+      ..publishMessage(
+        topic,
+        qualityOfService,
+        Uint8Buffer()..addAll((await content.byteBuffer).asUint8List()),
+        retain: form.retain ?? false,
+      )
+      ..disconnect();
+
+    return Content('application/octec-stream', const Stream<List<int>>.empty());
```

Comment:
Also, maybe the content-type should be a constant somewhere? That way typos like these would be avoided.
```suggestion
    return Content('application/octet-stream', const Stream<List<int>>.empty());
```

Reply:
Good catch :) Added a constant in acfd58e03642449890368e5bb9bd8a9322de0afa